### DISCUSSION TITLE
Accepting files and data in the same request -- there is a better way

### DISCUSSION THREAD

**Discussion by kvas-it** [fastapi/fastapi#8757] [Questions]
I posted this as a comment into [this ticket](https://github.com/tiangolo/fastapi/issues/3132#issuecomment-875725103) but there was no reply there, so I'll try my luck by creating a new ticket.

I have a route like this:
```
@app.post('/data_and_files')
def data_and_files(data: DataSchema = Form(...), files: List[UploadFile] = File([])):
    # here data would be parsed DataSchema and files is a list of files.
```

FastAPI docs would say that this is not going to work "because of limitations of HTTP protocol" but FastAPI will actually render a form for submitting this request under `/docs`.  The form will produce a request like this (without the `application/json` content type in the first part -- I added that to make it even more kosher):

```--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="data"
Content-Type: application/json

{"name": "John Doe", "age": 42}
--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="files"; filename="file1.txt"
Content-Type: text/plain

content 1
--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="files"; filename="file2.html"
Content-Type: text/html

<html/>
--ecbe7a002f5cb793f8ce4897649593f6--
```

FastAPI will refuse to parse the first part, because "it's not a dictionary" and I get HTTP 422. The request above is a perfectly valid HTTP request that does exactly the thing that the docs say is impossible. FastAPI could parse the `data` part into `DataSchema` as I have specified if only it realized that it should be JSON-parsed first.

For now I'm forced to change my code like this:

```
@app.post('/data_and_files')
def data_and_files(data: str = Form(...), files: List[UploadFile] = File([])):
    data = DataSchema.parse_raw(data)
    # here data would be parsed DataSchema and files is a list of files.
```

This works, but it has two disadvantages:
1.  I have to convert parsing errors into `HTTPException`s myself.
2. `/docs` shows `data` as a string, so I have to remember myself what to put into it (and when I tell my friends how to do it they complain that my API is hard to use).

All of this is not the end of the world, but it would be even better if my original code worked. Seems like FastAPI would only need to notice that it's trying to cast a string into a Pydantic model and use `DataSchema.parse_raw()` to do that. Wouldn't this be better than saying that HTTP doesn't let us do it?

**Comment by Mause** [comment]
Is there a reason why you're passing json in a form field, and not using discrete form fields for your data? The swagger UI would show all the form fields with the proper data that way

**Comment by kvas-it** [comment]
Thanks for the suggestion. Indeed, this is a way to get around this issue. I think at this point I prefer to have all the data in one JSON (even at the cost of not having adequate swagger UI support), but perhaps you could say that it didn't have to be this way.

What I like about the current structure of the API is that it's simple and coherent. In all the calls one JSON structure with all the fields goes in (like in [this tutorial](https://fastapi.tiangolo.com/tutorial/body/)), another JSON structure goes out and both of them are defined as Pydatic schemas. After I define the schemas, the web API is more or less autogenerated, so there's much less code repetition.

If it was just about the calls that have file uploads, perhaps refactoring the code to have each field of incoming data structure in a separate form field would make sense (although some of them have further structure inside, so I would run into the same problems anyway (and I can't flatten the structure out, because some fields are lists)). However, what I have is that most of the calls don't have attachments, so my approach works perfectly for them, and I don't want to lose the API uniformity. I also don't want to lose autogeneration.

So to summarize, I suppose you could say that I'm using FastAPI wrong, but the way I see it is: this approach works for my project, it gives me the API that I like, the code is nice and schema-driven with little boilerplate. There are just two inconveniences: I have to do input validation myself (no helpful type annotations) and I have a less helpful Swagger UI for the calls that have file attachments -- all of this is no big deal. A seemingly small change to FastAPI would make things nicer for me and other people like me, so I thought I would propose it.

**Comment by adriangb** [comment]
You are totally right, this is 100% valid HTTP.
Which means this is literally a FastAPI limitation.

Now, perhaps FastAPI doesn't want to support this, or maybe it's a bad idea, regardless the docs are definitely incorrect and should be updated

**Comment by markkorput** [comment]
> Is there a reason why you're passing json in a form field, and not using discrete form fields for your data? The swagger UI would show all the form fields with the proper data that way

WELL pydantic models don't seem to work with form data (which is surprising on its own, but the lack of information about this in the docs is even more surprising). There are `as_form` workaround [out there](https://stackoverflow.com/questions/60127234/how-to-use-a-pydantic-model-with-form-data-in-fastapi) to make Pydantic models work with form data, but I haven't been able to get them to work with complex/nested models.

My workaround for now is to turn @kvas-it's solution into a re-usable dependency (inspired by the aformentioned [as_form](https://stackoverflow.com/questions/60127234/how-to-use-a-pydantic-model-with-form-data-in-fastapi) solution). 

```
import inspect
from typing import Any, Callable, Optional, Type

from fastapi import Form
from pydantic import BaseModel


def as_form_json(cls: Type[BaseModel], attr_name: str = "data", default: Optional[str] = None) -> Callable[..., BaseModel]:
    """returns an alternative constructor for the given pydantic model that
    can be used as a fastapi dependency when the Model's is expected as form-data with a single
    string-attribute which contains json-encoded data for the model.

    Usage:
    ```py
        @router.post("/products")
        async def create_product(params: ProductParams = Depends(as_form_json(ProductParams))):
            pass
    ```
    """

    def as_form_func(**kwargs: Any) -> BaseModel:
        return cls.parse_raw(list(kwargs.values())[0])

    new_parameters = [
        inspect.Parameter(
            attr_name,
            inspect.Parameter.POSITIONAL_ONLY,
            default=default or Form(...),
            # annotation=model_field.outer_type_,
        )
    ]

    # modify method signature so it has an argument with the name specified by `attr_name`
    sig = inspect.signature(as_form_func)
    sig = sig.replace(parameters=new_parameters)
    as_form_func.__signature__ = sig  # type: ignore
    return as_form_func
```
